content/blog/2020-05-22-rt-template/index.md

```diff
@@ -0,0 +1,365 @@
+
+Therefore, functions offer a performant tradeoff for an accelerator. But in terms of productive 
+implementation, it is very program specific. The runtime library and the high-level 
+programmer needs to follow ad-hoc set of rules to use this template, even though it is
+possible to add some checks in the llvm pass with some engineering.
+
+## Replacing basic blocks
+Replacing a basic block can be a middle ground between these two granularities.
+A basic block can be considered as a collection of instructions, for the generation
+of interfaces. Since the interfaces for an instruction are clear, basic blocks will be an extension.
+It can be seen as a part of a function for the purposes of granularity. So a (loosely coupled) accelerator
+can work at a higher granularity than an instruction.
+
+```c
+    /* create the function call from runtime library */
+    LLVMContext& Ctx = F.getContext();
+    FunctionCallee rtFunc;
+    rtFunc = F.getParent()->getOrInsertFunction(
+      "bb_replace", 
+      Type::getInt32Ty(Ctx),
+      Type::getInt32Ty(Ctx)
+    );
+    // function call name and argument types have to be known
+
+    /* Add runtime basic block selection here */
+    
+    /* End of selection algorithm */
+    
+    /* Select a basic block */
+    BasicBlock* workList = NULL; // collects replaced basic block within iterator
+    /* mark basic blocks for deletion */
+    for (auto& B : F) {
+        for (auto& I : B) {
+            workList = &B;
+        }
+    }
+    
+    /* collect information on the basic block */
+    BasicBlock* predBB = workList->getSinglePredecessor();
+    BasicBlock* succBB = workList->getSingleSuccessor();
+    
+    /* delete basic block being replaced */
+    if (workList != NULL) {
+        workList->eraseFromParent();
+    }
+    
+    /* create new basic block */ 
+    BasicBlock *pb = BasicBlock::Create(Ctx, "simple", &F);
+    
+    /* create the runtime function call */
+    IRBuilder<> builder(pb);
+    Constant *i32_configs = ConstantInt::get(Type::getInt32Ty(Ctx), /*configuration*/, true);
+    Value* args[] = {i32_configs};
+    Value* ret =  builder.CreateCall(rtFunc, args);
+    
+    /* create a jump to next block */
+    Instruction *brInst = BranchInst::Create(succBB, pb);
+    
+    /* update jump from previous block */
+    Instruction *brPred = predBB->getTerminator();
+    brPred->setSuccessor(0, pb);
+```
+However, this also conflates challenges from both prior levels of abstraction.
+Unlike a single instruction, it's highly program dependent how a basic block interacts with
+rest of the program. Unlike a function, it is not clear what those program specific interfaces
+are, and has to be created at an intermediate-representation level. To make matters worse,
+these interfaces are intermediate-repesentation specific. If you target a specific optimization
+(say -O1), your pass will probably need updating for a different optimization.
```

Comment:
This needs explanation. Why would your pass need to change at different optimization levels?

Reply:
I removed the optimization level with code pattern.